File: 2022/day07/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Solution {
    public static void main(String[] args) throws IOException {
        Path inputPath = Paths.get(System.getProperty("user.dir")).resolve("../input.txt");
        List<String> lines = Files.readAllLines(inputPath);

        Map<String, Long> dirSizes = parseFilesystem(lines);

        System.out.println("Part 1: " + part1(dirSizes));
        System.out.println("Part 2: " + part2(dirSizes));
    }

    private static Map<String, Long> parseFilesystem(List<String> lines) {
        List<String> path = new ArrayList<>();
        Map<String, Long> dirSizes = new HashMap<>();

        for (String line : lines) {
            if (line.startsWith("$ cd")) {
                String target = line.substring(5);
                if (target.equals("/")) {
                    path.clear();
                    path.add("/");
                } else if (target.equals("..")) {
                    path.remove(path.size() - 1);
                } else {
                    path.add(target);
                }
            } else if (line.startsWith("$ ls") || line.startsWith("dir ")) {
                continue;
            } else {
                // It's a file with size
                String[] parts = line.split(" ");
                long size = Long.parseLong(parts[0]);

                // Add size to current directory and all parent directories
                for (int i = 0; i < path.size(); i++) {
                    String dirPath = String.join("/", path.subList(0, i + 1));
                    if (dirPath.isEmpty()) {
                        dirPath = "/";
                    }
                    dirSizes.merge(dirPath, size, Long::sum);
                }
            }
        }

        return dirSizes;
    }

    private static long part1(Map<String, Long> dirSizes) {
        return dirSizes.values().stream()
            .filter(size -> size <= 100000)
            .mapToLong(Long::longValue)
            .sum();
    }

    private static long part2(Map<String, Long> dirSizes) {
        long totalSpace = 70000000;
        long neededSpace = 30000000;
        long usedSpace = dirSizes.get("/");
        long freeSpace = totalSpace - usedSpace;
        long needToFree = neededSpace - freeSpace;

        return dirSizes.values().stream()
            .filter(size -> size >= needToFree)
            .mapToLong(Long::longValue)
            .min()
            .orElse(0);
    }
}

File: 2022/day07/rust/solution.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;

fn parse_filesystem(lines: &[&str]) -> HashMap<String, u64> {
    let mut path: Vec<&str> = Vec::new();
    let mut dir_sizes: HashMap<String, u64> = HashMap::new();

    for line in lines {
        if line.starts_with("$ cd") {
            let target = &line[5..];
            match target {
                "/" => {
                    path.clear();
                    path.push("/");
                }
                ".." => {
                    path.pop();
                }
                _ => {
                    path.push(target);
                }
            }
        } else if line.starts_with("$ ls") || line.starts_with("dir ") {
            continue;
        } else {
            // It's a file with size
            let size: u64 = line.split_whitespace().next().unwrap().parse().unwrap();
            // Add size to current directory and all parent directories
            for i in 0..path.len() {
                let dir_path = if i == 0 {
                    "/".to_string()
                } else {
                    path[..=i].join("/")
                };
                *dir_sizes.entry(dir_path).or_insert(0) += size;
            }
        }
    }

    dir_sizes
}

fn part1(dir_sizes: &HashMap<String, u64>) -> u64 {
    dir_sizes.values().filter(|&&size| size <= 100000).sum()
}

fn part2(dir_sizes: &HashMap<String, u64>) -> u64 {
    let total_space: u64 = 70000000;
    let needed_space: u64 = 30000000;
    let used_space = dir_sizes.get("/").unwrap();
    let free_space = total_space - used_space;
    let need_to_free = needed_space - free_space;

    *dir_sizes
        .values()
        .filter(|&&size| size >= need_to_free)
        .min()
        .unwrap()
}

fn main() {
    let exe_path = std::env::current_exe().unwrap();
    let exe_dir = exe_path.parent().unwrap();
    let input_path = exe_dir.join("../input.txt");

    let input = if Path::new(&input_path).exists() {
        fs::read_to_string(&input_path).unwrap()
    } else {
        fs::read_to_string("../input.txt").unwrap()
    };

    let lines: Vec<&str> = input.trim().lines().collect();
    let dir_sizes = parse_filesystem(&lines);

    println!("Part 1: {}", part1(&dir_sizes));
    println!("Part 2: {}", part2(&dir_sizes));
}
